sklearn_numba_dpex/kmeans/kernels.py

```diff
@@ -0,0 +1,807 @@
+
+
+@lru_cache
+def make_copyto_2d_kernel(size0, size1, work_group_size):
+    global_size = math.ceil(size1 / work_group_size) * work_group_size
+
+    # Optimized for C-contiguous array and for
+    # size1 >> preferred_work_group_size_multiple
+    @dpex.kernel
+    def copyto_kernel(input_array, target_array):
+        col_idx = dpex.get_global_id(0)
+
+        if col_idx >= size1:
+            return
+
+        for row_idx in range(size0):
+            target_array[row_idx, col_idx] = input_array[row_idx, col_idx]
+
+    return copyto_kernel[global_size, work_group_size]
+
+
+@lru_cache
+def make_broadcast_division_1d_2d_kernel(size0, size1, work_group_size):
+    global_size = math.ceil(size1 / work_group_size) * work_group_size
+
+    # NB: inplace. # Optimized for C-contiguous array and for
+    # size1 >> preferred_work_group_size_multiple
+    @dpex.kernel
+    def broadcast_division(dividend_array, divisor_vector):
+        col_idx = dpex.get_global_id(0)
+
+        if col_idx >= size1:
+            return
+
+        divisor = divisor_vector[col_idx]
+
+        for row_idx in range(size0):
+            dividend_array[row_idx, col_idx] = (
+                dividend_array[row_idx, col_idx] / divisor
+            )
+
+    return broadcast_division[global_size, work_group_size]
+
+
+@lru_cache
+def make_half_l2_norm_2d_axis0_kernel(size0, size1, work_group_size, dtype):
+    global_size = math.ceil(size1 / work_group_size) * work_group_size
+    zero = dtype(0.0)
+    two = dtype(2.0)
+
+    # Optimized for C-contiguous array and for
+    # size1 >> preferred_work_group_size_multiple
+    # fmt: off
+    @dpex.kernel
+    def half_l2_norm(
+            data,     # IN        (size0, size1)
+            result    # OUT       (size1,)
+            ):    
+    # fmt: on
+        col_idx = dpex.get_global_id(0)
+
+        if col_idx >= size1:
+            return
+
+        l2_norm = zero
+
+        for row_idx in range(size0):
+            item = data[row_idx, col_idx]
+            l2_norm += item * item
+
+        result[col_idx] = l2_norm / two
+
+    return half_l2_norm[global_size, work_group_size]
+
+
+@lru_cache
+def make_sum_reduction_1d_kernel(size, work_group_size, device, dtype):
+    """numba_dpex does not provide tools such as `cuda.reduce` so we implement from
+    scratch a reduction strategy. The strategy relies on the commutativity of the
+    operation used for the reduction, thus allowing to reduce the input in any order.
+
+    The strategy consists in performing local reductions in each work group using local
+    memory where each work item combine two values, thus halving the number of values,
+    and the number of active work items. At each iteration the work items are discarded
+    in a bracket manner. The work items with the greatest ids are discarded first, and
+    we rely on the fact that the remaining work items are adjacents to optimize the RW
+    operations.
```

Comment:
> elegant

It is, and a similar strategy can be applied to reductions with all kind of operators, and even for more complex reductions such as the argpartition we've talked about, but honestly if the performance mattered I'm not sure it would be competitive.